consensus/hotstuff_replica.go

```diff
@@ -195,20 +208,40 @@ func (m *ConsensusModule) validateProposal(msg *typesCons.HotstuffMessage) error
 	}
 
 	// Safety: check the hash of the locked QC
-	// TODO(olshansky): Extend implementation to adopt `ExtendsFrom` as described in the Hotstuff whitepaper.
-	if protoHash(lockedQC.Block) == protoHash(justifyQC.Block) { // && lockedQC.Block.ExtendsFrom(justifyQC.Block)
+	// The equivalent of `lockedQC.Block.ExtendsFrom(justifyQC.Block)` in the hotstuff whitepaper is done in `applyBlock` below.
+	if protoHash(lockedQC.GetBlock()) == protoHash(justifyQC.Block) {
 		m.nodeLog(typesCons.ProposalBlockExtends)
 		return nil
 	}
 
-	// Liveness: node is locked on a QC from the past. [TODO]: Do we want to set `m.LockedQC = nil` here or something else?
+	// Liveness: node is locked on a QC from the past.
+	// DISCUSS: Where should additional logic be added to unlock the node?
 	if justifyQC.Height > lockedQC.Height || (justifyQC.Height == lockedQC.Height && justifyQC.Round > lockedQC.Round) {
 		return typesCons.ErrNodeIsLockedOnPastQC
 	}
 
 	return typesCons.ErrUnhandledProposalCase
 }
 
+// This helper applies the block metadata to the utility & persistence layers
+func (m *ConsensusModule) applyBlock(block *typesCons.Block) error {
+	// TECHDEBT: Retrieve this from persistence
+	lastByzValidators := make([][]byte, 0)
+
+	// Apply all the transactions in the block and get the appHash
+	appHash, err := m.UtilityContext.ApplyBlock(int64(m.Height), block.BlockHeader.ProposerAddress, block.Transactions, lastByzValidators)
```

Comment:
I have a couple of "problems"/questions with this and other instances where we access `m.UtilityContext`:

1. `m.UtilityContext` -could- be `nil` and we would simply panic here.

   I understand that we call `refreshUtilityContext` when we start a new round, I was wondering if we can just safely rely on that. 

2. If a replica receives a `Prepare` message that passes validations and the handler is invoked down the line, we would find ourselves here trying to apply a block but with which `m.UtilityContext`? What if it was started with a different height and what if is `nil` instead?

Reply:
I see what you mean.

1. Re nil - I think we can easily check everywhere, but it's just not pretty. I'm holding off on doing that until we make a decision.

2. Regarding "race conditions" - we can definitely get into weird "stateful" situations where we have contexts and messages at different heights.

3. I'm thinking of modifying the UtilityModule interface to have a "InitializeContext(height)" or "PrepareContext(height)" functions to simplify everything and make it stateless. The business logic on block validation moves entirely to the utility context. However, I'd want to discuss that with @andrewnguyen22 prior.

Reply:
Since it requires some thinkering / discussion I'd say that for now it's just best to just track it in an issue and move forward.